src/sage/schemes/weighted_projective/weighted_projective_space.py

```diff
@@ -0,0 +1,418 @@
+    if isinstance(weights, (int, Integer)):
+        return ProjectiveSpace(weights, R=R, names=names)
+    elif isinstance(weights, (list, tuple)):
+        # Make it hashable
+        weights = tuple(map(Integer, weights))
+        if any(w <= 0 for w in weights):
+            raise TypeError(
+                f"weights(={weights}) should only consist of positive integers"
+            )
+    else:
+        raise TypeError(f"weights={weights} must be an integer, a list or a tuple")
+
+    if names is None:
+        names = "x"
+
+    # TODO: Specialise implementation to projective spaces over non-rings.
+    if R in _CommRings:
+        return WeightedProjectiveSpace_ring(weights, R=R, names=names)
+
+    raise TypeError(f"R (={R}) must be a commutative ring")
+
+
+class WeightedProjectiveSpace_ring(UniqueRepresentation, AmbientSpace):
+    @staticmethod
+    def __classcall__(cls, weights: tuple[Integer], R=ZZ, names=None):
+        # __classcall_ is the "preprocessing" step for UniqueRepresentation
+        # see docs of CachedRepresentation
+        # weights should be a tuple, also because it should be hashable
+        if not isinstance(weights, tuple):
+            raise TypeError(
+                f"weights(={weights}) is not a tuple. Please use the"
+                "`WeightedProjectiveSpace` constructor"
+            )
+
+        normalized_names = normalize_names(len(weights), names)
+        return super().__classcall__(cls, weights, R, normalized_names)
+
+    def __init__(self, weights: tuple[Integer], R=ZZ, names=None):
+        """
+        Initialization function.
+
+        EXAMPLES::
+
+            sage: WeightedProjectiveSpace(Zp(5), [1, 3, 1], 'y')                        # needs sage.rings.padics
+            Weighted Projective Space of dimension 2 with weights (1, 3, 1) over 5-adic Ring with
+            capped relative precision 20
+            sage: WeightedProjectiveSpace(QQ, 5, 'y')
+            Projective Space of dimension 5 over Rational Field
+            sage: _ is ProjectiveSpace(QQ, 5, 'y')
+            True
+        """
+        AmbientSpace.__init__(self, len(weights) - 1, R)
+        self._weights = weights
+        self._assign_names(names)
+
+    def weights(self) -> tuple[Integer]:
+        """
+        Return the tuple of weights of this weighted projective space.
+
+        EXAMPLES::
+
+            sage: WeightedProjectiveSpace(QQ, [1, 3, 1]).weights()
+            (1, 3, 1)
+        """
+        return self._weights
+
+    def ngens(self) -> Integer:
+        """
+        Return the number of generators of this weighted projective space.
+
+        This is the number of variables in the coordinate ring of ``self``.
+
+        EXAMPLES::
+
+            sage: WeightedProjectiveSpace(QQ, [1, 3, 1]).ngens()
+            3
+            sage: WeightedProjectiveSpace(ZZ, 5).ngens()
+            6
+        """
+        return self.dimension_relative() + 1
```

Comment:
```suggestion
        return self.dimension_relative() + ZZ.one()
```
Avoid `int` to `Integer` coercion. This isn't a performance-critical method but since this file is importing `ZZ` anyway might as well.

Reply:
but it's ugly... I've a better idea, see https://github.com/sagemath/sage/pull/41178

Reply:
That is a better idea.